File: Longest Substring Without Repeating Characters/Longest_Substring_Without_Repeating_Characters.cpp
#include <iostream>
#include <unordered_map>
#include <string>
using namespace std;
class Solution
{
public:
    int lengthOfLongestSubstring(string s)
    {
        unordered_map<char, int> hash;
        int longest = 0;
        int start = 0;
        for(int i=0; i<s.length(); i++)
        {
            if(hash.find(s[i]) != hash.end() && hash[s[i]] >= start)
            {
                start = hash[s[i]] +1;
            }
            else
                longest = max(longest, i-start+1);    
            hash[s[i]] = i;
        }
        cout << longest << endl;
        return longest;
    }
};

int main()
{
    string s("abcabcbb");
    Solution a;
    a.lengthOfLongestSubstring(s);
    return 0;
}

File: Longest Substring Without Repeating Characters/Longest_Substring_Without_Repeating_Characters.py
# https://leetcode.com/problems/longest-substring-without-repeating-characters/
'''
Given a string, find the length of the longest substring without repeating characters.

Example 1:

Input: "abcabcbb"
Output: 3 
Explanation: The answer is "abc", with the length of 3. 
Example 2:

Input: "bbbbb"
Output: 1
Explanation: The answer is "b", with the length of 1.
Example 3:

Input: "pwwkew"
Output: 3
Explanation: The answer is "wke", with the length of 3. 
             Note that the answer must be a substring, "pwke" is a subsequence and not a substring.
'''

class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        last_seen = {}
        start = 0
        longest = 0

        for i, c in enumerate(s):
            if c in last_seen and last_seen[c] >= start:
                start = last_seen[c] + 1
            else:
                longest = max(longest, i - start +1)
            last_seen[c] = i
        return longest

if __name__ == "__main__":
    s = Solution()
    s.lengthOfLongestSubstring('abcabcbb')